node_manager.py

```diff
@@ -0,0 +1,134 @@
+class NodeManager():
+
+    def __init__(self, plugin_player):
+        self.plugin_player = plugin_player
+    
+    def link_input(self, instance):
+        #get info from instance id
+        output_node_number = int(instance.id.split()[0])
+        output_number = int(instance.id.split()[1])
+        input_node_number = int(instance.id.split()[2])
+        input_number = int(instance.id.split()[3])
+        input_node = self.plugin_player.nodes[input_node_number]
+        input_key = list(input_node.input_dict.keys())[input_number]
+
+        #check if input was previously set and wasn't a custom value
+        if input_node.input_map[input_number] and input_node.input_map[input_number][0] != -1:
+            #find previous output
+            previous_output_node_number = input_node.input_map[input_number][0]
+            previous_output_number = input_node.input_map[input_number][1]
+            previous_output_node = self.plugin_player.nodes[previous_output_node_number]
+            previous_output_key = list(previous_output_node.output_dict.keys())[previous_output_number]
+            
+            #delete old lines connecting to the input node
+            for line_set in input_node.module_widget.incoming_lines:
+                if line_set[1] == (previous_output_node_number, previous_output_number):
+                    input_node.module_widget.incoming_lines.remove(line_set)
+                    self.plugin_player.root.ids.right_section.ids.tree_section.canvas.before.remove(line_set[0])
+
+            #delete old lines connecting to the output node
+            for line_set in previous_output_node.module_widget.outgoing_lines:
+                if line_set[1] == (input_node_number, input_number):
+                    previous_output_node.module_widget.outgoing_lines.remove(line_set)
+                    self.plugin_player.root.ids.right_section.ids.tree_section.canvas.before.remove(line_set[0])
+
+            #remove the previous output edge
+            previous_output_node.output_map[previous_output_number].remove((input_node_number, input_number))
+            self.plugin_player.addMessage(f"Removed input {input_key} of Node {input_node_number} to {previous_output_key} from Node {previous_output_node_number}")
+        
+        #load in a custom input 
+        if output_node_number == -1:
+
+            #grab remaining values
+            input_node = self.plugin_player.nodes[input_node_number]
+            input_key = list(input_node.input_dict.keys())[input_number]
+
+            #grab input from currently opened popup entry
+            custom_declaration = input_node.custom_declaration_widget.text
+
+            #try to assign the variable
+            #may be a security risk as it evaluates injected code
+            try:
+                custom_input = eval(custom_declaration, sys.modules['__main__'].__dict__)
```

Comment:
```suggestion
            #try to assign the variable
            #may be a security risk as it evaluates injected code
            # TODO: find eval alternative
            try:
                custom_input = eval(custom_declaration, sys.modules['__main__'].__dict__)
```
Just to have an explicit note.